src/merkle/simple_smt/mod.rs

```diff
@@ -87,17 +87,10 @@ impl SimpleSmt {
             return Err(MerkleError::InvalidNumEntries(max, entries.len()));
         }
 
-        // append leaves to the tree returning an error if a duplicate entry for the same key
-        // is found
-        let mut empty_entries = BTreeSet::new();
+        let mut seen = BTreeSet::new();
         for (key, value) in entries {
-            let old_value = tree.update_leaf(key, value)?;
-            if old_value != Self::EMPTY_VALUE || empty_entries.contains(&key) {
-                return Err(MerkleError::DuplicateValuesForIndex(key));
-            }
-            // if we've processed an empty entry, add the key to the set of empty entry keys, and
-            // if this key was already in the set, return an error
-            if value == Self::EMPTY_VALUE && !empty_entries.insert(key) {
+            tree.update_leaf(key, value)?;
+            if !seen.insert(key) {
                 return Err(MerkleError::DuplicateValuesForIndex(key));
             }
```

Comment:
While this implementation is simpler, it would be somewhat less performant as we need to built a `BTreeSet` of size $n$ (for $n$ leaves). Is the idea that we don't care much about performance for now and can optimize this method (and the overall data structure) at a later point?

Reply:
changed it to track only empty word values.